### PR TITLE
UTF-8 target

### PR DESCRIPTION
**Pull request by reuben**
This is the UTF-8 target code. The alphabet size is 255 since we never see a zero byte, so I shift values by -1. With the CTC blank the final layer width is a round 256 units. I added a `--utf8` flag that enables UTF-8 training. Due to the recent changes to embed the alphabet in the exported model, changing from `Alphabet` to `UTF8Alphabet` is pretty much the only thing you have to do for UTF-8 training (this is done in `util/config.py` when initializing `Config.alphabet`). I've also made it so that in UTF-8 mode test report will sort samples by CER rather than WER. This is just for convenience since the main use of UTF-8 mode right now is for Mandarin models.

On the decoder side, I replaced the old code paths for character based LMs with UTF-8 based character language models. The idea is similar, but because we're predicting UTF-8 bytes directly, the logic for when to score a beam becomes a bit of a hybrid between word and character modes. Each "word" is a sequence of UTF-8 bytes that encode a single Unicode codepoint, and we score the beam when the the codepoint boundary is reached, rather when a space character is seen.

### CODE REVIEW

File: native_client/ctcdecode/path_trie.cpp [hunk 1]
```diff
@@ -111,25 +111,68 @@ PathTrie* PathTrie::get_path_trie(int new_char, int new_timestep, float cur_log_
...
+PathTrie* PathTrie::get_prev_grapheme(std::vector<int>& output,
+                                      std::vector<int>& timesteps)
+{
+  PathTrie* stop = this;
+  if (character == ROOT_) {
+    return stop;
+  }
+  // Recursive call: recurse back until stop condition, then append data in
+  // correct order as we walk back down the stack in the lines below.
+  //FIXME: use Alphabet instead of hardcoding +1 here
```
Comment (tilmankamp) [inline native_client/ctcdecode/path_trie.cpp]: How about fixing this before merging?

File: native_client/ctcdecode/decoder_utils.cpp [hunk 2]
```diff
@@ -61,13 +46,18 @@ size_t get_utf8_str_len(const std::string &str) {
   return str_len;
 }
 
-std::vector<std::string> split_utf8_str(const std::string &str) {
+// Return weather a byte is a code point boundary (not a continuation byte).
+bool byte_is_codepoint_boundary(unsigned char c) {
+  // only continuation bytes have their most significant bits set to 10
+  return (c & 0xC0) != 0x80;
```
[the remark this answers is not in the document]
Comment (reuben) [reply native_client/ctcdecode/decoder_utils.cpp]: Will do.

File: native_client/ctcdecode/path_trie.cpp [hunk 3]
```diff
@@ -111,25 +111,68 @@ PathTrie* PathTrie::get_path_trie(int new_char, int new_timestep, float cur_log_
...
+    stop = parent->get_prev_grapheme(output, timesteps);
+  }
+  output.push_back(character);
+  timesteps.push_back(timestep);
+  return stop;
+}
+
+int PathTrie::distance_to_codepoint_boundary(unsigned char *first_byte)
+{
+  //FIXME: use Alphabet instead of hardcoding +1 here
```
[the remark this answers is not in the document]
Comment (reuben) [reply native_client/ctcdecode/path_trie.cpp]: The problem is that these functions are quite hot and going through the Alphabet currently adds a whole lot of work due to its high level API. My plan was to figure out a more lightweight solution to fix this.

**Comment by reuben** [comment]
Filed https://github.com/mozilla/DeepSpeech/issues/2526 for the follow-up.

**Pull request merged**